mailpoet/lib/Util/Security.php

```diff
@@ -31,11 +32,49 @@ public static function generateToken($action = 'mailpoet_token') {
     return WPFunctions::get()->wpCreateNonce($action);
   }
 
+  public static function generateTokenForUniqueAction(string $action) {
+    return self::generateToken(self::UNIQUE_ACTION_PREFIX . $action);
+  }
+
+  public static function getUniqueAction(string $action) {
+    return self::UNIQUE_ACTION_PREFIX . $action;
+  }
+
+  public static function isActionUnique(string $action): bool {
+    return strpos($action, self::UNIQUE_ACTION_PREFIX) === 0;
+  }
+
+  public static function checkTokenForUniqueAction(string $token, string $action): bool {
+    return self::checkToken($token, self::getUniqueAction($action));
+  }
+
+  public static function checkToken(string $token, string $action = 'mailpoet_token'): bool {
+    $isNonceValid = WPFunctions::get()->wpVerifyNonce($token, $action);
+
+    if (!$isNonceValid || !self::isActionUnique($action)) {
+      return $isNonceValid;
+    }
+
+    // Nonces that were generated for a unique action should truly only ever be used once
```

Comment:
This next part may be overkill, or there may be a more elegant way of handling it, but I really wanted to make sure that these security tokens were truly a single time use.

I was a little worried that this could result in tons of transients cluttering up users' options tables, but then I found that WordPress has a daily cron task to delete expired transients.

Reply:
Yeah, it seems to be a bit unnecessary to me. The reason is, in other places, WP doesn't do that either (nonces are not one-time use) so us doing it doesn't seem to have that much sense to me. It's not possible to expire the nonces and https://codex.wordpress.org/WordPress_Nonces doesn't suggest attempting that.

So I'd stick to what WP recommends, as extending it to one-use would only add unnecessary complexity and possible performance issues on large sites. The security risk is questionable — if an attacker manages to get a nonce, they'd still need your session if I'm not mistaken.

Reply:
I agree that this is probably a bad solution since it's basically changing how nonces work, which could have unintended side effects and definitely adds more complexity than it's worth, especially since the issue itself doesn't seem valid as you point out in another comment.